packages/glob/src/internal-pattern.ts

```diff
@@ -0,0 +1,227 @@
+        .replace(/\*/g, '[*]') // escape '*'
+      pattern = pathHelper.ensureRooted(root, pattern)
+    }
+
+    // Normalize slashes
+    pattern = pathHelper.normalizeSeparators(pattern)
+
+    // Trailing slash indicates the pattern should only match directories, not regular files
+    this.trailingSlash = pathHelper
+      .normalizeSeparators(pattern)
+      .endsWith(path.sep)
+    pattern = pathHelper.safeTrimTrailingSeparator(pattern)
+
+    // Set search path
+    // Set root regexp
+    this.searchPath = (undefined as unknown) as string
+    this.rootRegExp = (undefined as unknown) as RegExp
+    this.initializePaths(pattern)
+
+    // Create minimatch
+    const minimatchOptions: IMinimatchOptions = {
+      dot: true,
+      nobrace: true,
+      nocase: IS_WINDOWS,
+      nocomment: true,
+      noext: true,
+      nonegate: true
+    }
+    pattern = IS_WINDOWS ? pattern.replace(/\\/g, '/') : pattern
+    this.minimatch = new Minimatch(pattern, minimatchOptions)
+  }
+
+  /**
+   * Matches the pattern against the specified path
+   */
+  match(itemPath: string): MatchResult {
+    if (this.comment) {
+      return MatchResult.None
+    }
+
+    if (this.minimatch.match(itemPath)) {
+      return this.trailingSlash ? MatchResult.Directory : MatchResult.All
+    }
+
+    return MatchResult.None
+  }
+
+  /**
+   * Indicates whether the pattern may match descendants of the specified path
+   */
+  partialMatch(itemPath: string): boolean {
+    if (this.comment) {
+      return false
+    }
+
+    // matchOne does not handle root path correctly
+    if (pathHelper.dirname(itemPath) === itemPath) {
+      return this.rootRegExp.test(itemPath)
+    }
+
+    return this.minimatch.matchOne(
+      itemPath.split(/\/+/),
+      this.minimatch.set[0],
+      true
+    )
+  }
+
+  /**
+   * Initializes the search path and root regexp
+   */
+  private initializePaths(pattern: string): void {
+    // Parse the pattern as a path
+    const patternPath = new Path(pattern)
+
+    // Build the search path
+    this.searchPath = ''
+    for (const patternSegment of patternPath.segments) {
+      // Convert
+      const literal = this.convertToLiteral(patternSegment)
+      if (!literal) {
+        break
+      }
+
+      // Append slash
+      // Note, this is OK because Pattern.ctor() asserts the path is not like C: or C:foo
+      if (this.searchPath && !this.searchPath.endsWith(path.sep)) {
+        this.searchPath += path.sep
+      }
+
+      // Append literal segment
+      this.searchPath += literal
+    }
+
+    // Store the root segment (required when determining partial match)
+    const rootSegment = new Path(pattern).segments[0]
+    this.rootRegExp = new RegExp(
+      this.regExpEscape(rootSegment),
+      IS_WINDOWS ? 'i' : ''
+    )
+
+    // Set the root regexp
+    this.rootRegExp = new RegExp(
+      this.regExpEscape(this.convertToLiteral(patternPath.segments[0]))
+    )
+  }
+
+  /**
```

Comment:
This function is required to accurately determine the search path. Can't simply use information from minimatch (whether a segment \"has magic\") because users sometimes need to use `[ ]` for escaping on Windows.